File: ios/Packages/PadelCore/Sources/PadelCore/Sync/MandoDeTanda.swift
import Foundation

/// El mando a distancia de las tandas de entrenamiento: el móvil ordena, el reloj obedece.
///
/// Existe porque la tanda de datos casi nunca la graba quien lleva el reloj. Lo normal es
/// ponérselo a otra persona y dirigirla desde fuera —"ahora treinta derechas", "ahora
/// bandejas"— y con los botones solo en la muñeca hay que parar el ejercicio, quitarle el
/// reloj, cambiar el tipo y volver a empezar. Eso rompe la tanda y, peor, invita a grabar
/// menos tandas de las que hacen falta.
///
/// Las órdenes viajan por dos caminos y **nunca** por contexto de aplicación: un contexto
/// se reentrega al reconectar, y una orden reentregada arrancaría una tanda que nadie ha
/// pedido.
///
/// - Con el reloj a mano (su app en primer plano) va por `sendMessage`, que contesta al
///   instante con el estado entero.
/// - Con el reloj dormido va por `transferUserInfo`, que **encola** y despierta la app del
///   reloj en segundo plano. Cada elemento se entrega una sola vez, así que no hay riesgo
///   de reentrega, y el reloj devuelve su estado por el mismo camino.
///
/// El segundo camino es el que hace que el mando sirva para algo: mirar el móvil apaga la
/// pantalla del reloj, y exigir que estuviera despierto era exigirlo justo en el momento
/// en que no puede estarlo.
public enum AccionDeTanda: String, Codable, Sendable {
    /// No cambia nada: solo pregunta cómo va. Es lo que refresca el contador.
    case estado
    case iniciar
    case parar
    /// Manda al móvil lo que haya guardado.
    case enviar
}

/// Una orden del móvil al reloj.
public struct OrdenDeTanda: Codable, Sendable {
    public let accion: AccionDeTanda
    /// Tipo de golpe que se va a grabar. Solo lo mira `iniciar`.
    public let etiqueta: ShotType?
    /// Cuándo se dio la orden. Nil en órdenes de versiones viejas.
    public let creadoEpochMs: Int64?

    public init(accion: AccionDeTanda, etiqueta: ShotType? = nil, creadoEpochMs: Int64? = nil) {
        self.accion = accion
        self.etiqueta = etiqueta
        self.creadoEpochMs = creadoEpochMs
    }

    /// La clave del mensaje en WatchConnectivity. Una sola, compartida por los dos lados.
    public static let clave = "padel_mando_tanda"

    /// Una orden encolada puede tardar en llegar si el reloj estaba sin cobertura. Pasado
    /// este rato ya no se obedece: arrancar una tanda diez minutos después de pedirla, con
    /// el reloj otra vez en la muñeca de otro, es peor que no arrancarla.
    public static let maxAntiguedadMs: Int64 = 3 * 60_000

    /// ¿Sigue vigente esta orden? Las de `estado` siempre lo están: no cambian nada.
    public func vigente(ahoraEpochMs: Int64) -> Bool {
        if accion == .estado { return true }
        guard let creadoEpochMs else { return true }
        return ahoraEpochMs - creadoEpochMs <= Self.maxAntiguedadMs
    }
}

/// Lo que el reloj contesta a cualquier orden: su estado entero.
public struct EstadoDeTanda: Codable, Sendable, Equatable {
    public let grabando: Bool
    public let etiqueta: ShotType
    /// Golpeos capturados en la tanda en curso.
    public let capturadosEnTanda: Int
    /// Golpeos guardados en el reloj, de todas las tandas.
    public let guardadosEnTotal: Int
    public let kilobytes: Int
    /// Quién lleva el reloj ahora mismo, según los ajustes replicados.
    public let alias: String
    /// Su nivel técnico declarado, que es lo que ancla la escala. Nil = sin declarar.
    public let nivel: Int?
    /// El reloj avisa de que sin permiso de entreno la tanda puede cortarse al apagarse
    /// la pantalla. Devolver diez golpes de cincuenta en silencio sería peor que fallar.
    public let sensoresPuedenPararse: Bool
    /// Por qué la última orden no hizo lo que se le pidió, si es que no lo hizo. Nil =
    /// todo en orden. Sin esto, pulsar "Grabar" durante un partido deja un botón que
    /// parece roto en vez de una explicación.
    public let motivo: String?
    /// Los swings que el detector tiró en esta tanda, con el motivo. Es lo que convierte
    /// "no me coge la mitad de las bandejas" en un número con su umbral culpable al lado.
    public let descartes: DescartesDelDetector?
    /// Segundos de tanda grabados, si hay una en marcha. Es el contador que siempre
    /// avanza: la prueba de que se está guardando algo, vea el detector lo que vea.
    public let segundosDeTanda: Int?

    /// La clave con la que el reloj devuelve su estado cuando la orden vino encolada y
    /// no había a quién contestar en el momento.
    public static let clave = "padel_estado_tanda"

    public init(
        grabando: Bool,
        etiqueta: ShotType,
        capturadosEnTanda: Int,
        guardadosEnTotal: Int,
        kilobytes: Int,
        alias: String,
        nivel: Int?,
        sensoresPuedenPararse: Bool,
        motivo: String? = nil,
        descartes: DescartesDelDetector? = nil,
        segundosDeTanda: Int? = nil
    ) {
        self.grabando = grabando
        self.etiqueta = etiqueta
        self.capturadosEnTanda = capturadosEnTanda
        self.guardadosEnTotal = guardadosEnTotal
        self.kilobytes = kilobytes
        self.alias = alias
        self.nivel = nivel
        self.sensoresPuedenPararse = sensoresPuedenPararse
        self.motivo = motivo
        self.descartes = descartes
        self.segundosDeTanda = segundosDeTanda
    }
}

File: android/core/src/main/kotlin/com/risingpadel/core/sync/MandoDeTanda.kt
package com.risingpadel.core.sync

import com.risingpadel.core.detection.DescartesDelDetector
import com.risingpadel.core.model.ShotType
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * El mando a distancia de las tandas de entrenamiento: el móvil ordena, el reloj obedece.
 *
 * Existe porque la tanda de datos casi nunca la graba quien lleva el reloj. Lo normal es
 * ponérselo a otra persona y dirigirla desde fuera —"ahora treinta derechas", "ahora
 * bandejas"— y con los botones solo en la muñeca hay que parar el ejercicio, quitarle el
 * reloj, cambiar el tipo y volver a empezar. Eso rompe la tanda y, peor, invita a grabar
 * menos tandas de las que hacen falta.
 *
 * En Wear OS las órdenes van por `MessageClient`, que despierta el
 * `WearableListenerService` del reloj aunque su app esté cerrada y entrega cada mensaje
 * una sola vez. No van por `DataClient`: un data item se reentrega al reconectar, y una
 * orden reentregada arrancaría una tanda que nadie ha pedido — el mismo motivo por el
 * que en el lado Apple esto no viaja por contexto de aplicación.
 */
@Serializable
enum class AccionDeTanda {
    /** No cambia nada: solo pregunta cómo va. Es lo que refresca el contador. */
    @SerialName("estado")
    ESTADO,

    @SerialName("iniciar")
    INICIAR,

    @SerialName("parar")
    PARAR,

    /** Manda al móvil lo que haya guardado. */
    @SerialName("enviar")
    ENVIAR,
}

/** Una orden del móvil al reloj. */
@Serializable
data class OrdenDeTanda(
    val accion: AccionDeTanda,
    /** Tipo de golpe que se va a grabar. Solo lo mira `INICIAR`. */
    val etiqueta: ShotType? = null,
    /** Cuándo se dio la orden. Nil en órdenes de versiones viejas. */
    val creadoEpochMs: Long? = null,
) {
    /**
     * ¿Sigue vigente esta orden? Las de `ESTADO` siempre lo están: no cambian nada.
     */
    fun vigente(ahoraEpochMs: Long): Boolean {
        if (accion == AccionDeTanda.ESTADO) return true
        val creado = creadoEpochMs ?: return true
        return ahoraEpochMs - creado <= MAX_ANTIGUEDAD_MS
    }

    fun encode(): String = JSON.encodeToString(serializer(), this)

    companion object {
        /** La ruta del mensaje en el Data Layer. Una sola, compartida por los dos lados. */
        const val PATH = "/padel/mando-tanda"

        /**
         * Una orden puede tardar en llegar si el reloj estaba sin cobertura. Pasado este
         * rato ya no se obedece: arrancar una tanda diez minutos después de pedirla, con
         * el reloj otra vez en la muñeca de otro, es peor que no arrancarla.
         */
        const val MAX_ANTIGUEDAD_MS: Long = 3 * 60_000

        fun decode(json: String): OrdenDeTanda? =
            runCatching { JSON.decodeFromString(serializer(), json) }.getOrNull()
    }
}

/** Lo que el reloj contesta a cualquier orden: su estado entero. */
@Serializable
data class EstadoDeTanda(
    val grabando: Boolean,
    val etiqueta: ShotType,
    /** Golpeos capturados en la tanda en curso. */
    val capturadosEnTanda: Int,
    /** Golpeos guardados en el reloj, de todas las tandas. */
    val guardadosEnTotal: Int,
    val kilobytes: Int,
    /** Quién lleva el reloj ahora mismo, según los ajustes replicados. */
    val alias: String,
    /** Su nivel técnico declarado, que es lo que ancla la escala. Nil = sin declarar. */
    val nivel: Int? = null,
    /**
     * El reloj avisa de que sin permiso de entreno la tanda puede cortarse al apagarse la
     * pantalla. Devolver diez golpes de cincuenta en silencio sería peor que fallar.
     */
    val sensoresPuedenPararse: Boolean = false,
    /**
     * Por qué la última orden no hizo lo que se le pidió, si es que no lo hizo. Nil = todo
     * en orden. Sin esto, pulsar "Grabar" durante un partido deja un botón que parece roto
     * en vez de una explicación.
     */
    val motivo: String? = null,
    /**
     * Los swings que el detector tiró en esta tanda, con el motivo. Es lo que convierte
     * "no me coge la mitad de las bandejas" en un número con su umbral culpable al lado.
     */
    val descartes: DescartesDelDetector? = null,
    /**
     * Segundos de tanda grabados, si hay una en marcha. Es el contador que siempre
     * avanza: la prueba de que se está guardando algo, vea el detector lo que vea.
     */
    val segundosDeTanda: Int? = null,
) {
    fun encode(): String = JSON.encodeToString(serializer(), this)

    companion object {
        /** La ruta por la que el reloj devuelve su estado. */
        const val PATH = "/padel/estado-tanda"

        fun decode(json: String): EstadoDeTanda? =
            runCatching { JSON.decodeFromString(serializer(), json) }.getOrNull()
    }
}

/**
 * Tolerante con campos desconocidos a propósito: los dos lados se actualizan por
 * separado —el reloj por Play Store, el móvil también— y un mando que deja de funcionar
 * porque la otra mitad añadió un campo es un mando roto en pista.
 */
private val JSON = Json {
    ignoreUnknownKeys = true
    encodeDefaults = true
}
